src/Microsoft.Azure.WebJobs.Extensions.Kafka/Listeners/KafkaTopicScaler.cs

```diff
@@ -0,0 +1,247 @@
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.OfType<KafkaTriggerMetrics>().ToArray());
+        }
+
+        public ScaleStatus GetScaleStatus(ScaleStatusContext<KafkaTriggerMetrics> context)
+        {
+            return GetScaleStatusCore(context.WorkerCount, context.Metrics?.ToArray());
+        }
+
+        private ScaleStatus GetScaleStatusCore(int workerCount, KafkaTriggerMetrics[] metrics)
+        {
+            var status = new ScaleStatus
+            {
+                Vote = ScaleVote.None,
+            };
+
+            const int NumberOfSamplesToConsider = 5;
+
+            // At least 5 samples are required to make a scale decision for the rest of the checks.
+            if (metrics == null || metrics.Length < NumberOfSamplesToConsider)
+            {
+                return status;
+            }
+
+            var lastMetrics = metrics.Last();
+            long totalLag = lastMetrics.TotalLag;
+            long partitionCount = lastMetrics.PartitionCount;
+            long lagThreshold = 1000L;
+
+            //This case shouldn't ever happens as we do not scale out in case workerCount >= partitionCount
+            if (partitionCount > 0 && partitionCount < workerCount)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+
+                if (this.logger.IsEnabled(LogLevel.Information))
+                {
+                    this.logger.LogInformation("WorkerCount ({workerCount}) > PartitionCount ({partitionCount}). For topic {topicName}, for consumer group {consumerGroup}.", workerCount, partitionCount, this.topicName, this.consumerGroup);
+                    this.logger.LogInformation("Number of instances ({workerCount}) is too high relative to number of partitions ({partitionCount}). For topic {topicName}, for consumer group {consumerGroup}.", workerCount, partitionCount, this.topicName, this.consumerGroup);
+                }
+
+                return status;
+            }
+
+
+            // Check to see if the Kafka consumer has been empty for a while. Only if all metrics samples are empty do we scale down.
+            bool partitionIsIdle = metrics.All(p => p.TotalLag == 0);
+
+            if (partitionIsIdle)
+            {
+                status.Vote = ScaleVote.ScaleIn;
+                if (this.logger.IsEnabled(LogLevel.Information))
+                {
+                    this.logger.LogInformation("Topic '{topicName}', for consumer group {consumerGroup}' is idle.", this.topicName, this.consumerGroup);
+                }
+
+                return status;
+            }
+
+
+            // don't scale out beyond the number of partitions, Maintain a minimum ratio of 1 worker per 1,000 unprocessed messages.
+            if (metrics[0].TotalLag > 0 && totalLag / lagThreshold > partitionCount)
+            {
+                if (workerCount < partitionCount)
```

Comment:
Remove this check. You've verified this earlier above

Reply:
This check is need it to make sure this function will not vote up more than the number of partitions. the above condition checked for partitionCount < workerCount.